contracts/src/utils/structs/checkpoints.rs

```diff
@@ -0,0 +1,462 @@
+    }
+
+    /// Returns the value in the first (oldest) checkpoint with key
+    /// greater or equal than the search key, or zero if there is none.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    /// * `key` - Checkpoint's key to lookup.
+    pub fn lower_lookup(&self, key: U96) -> U160 {
+        let len = self.length();
+        let pos = self._lower_binary_lookup(key, U256::ZERO, len);
+        if pos == len {
+            U160::ZERO
+        } else {
+            self._access(pos)._value.get()
+        }
+    }
+
+    /// Returns the value in the last (most recent) checkpoint with key
+    /// lower or equal than the search key, or zero if there is none.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    /// * `key` - Checkpoint's key to lookup.
+    pub fn upper_lookup(&self, key: U96) -> U160 {
+        let len = self.length();
+        let pos = self._upper_binary_lookup(key, U256::ZERO, len);
+        if pos == U256::ZERO {
+            U160::ZERO
+        } else {
+            self._access(pos - uint!(1_U256))._value.get()
+        }
+    }
+
+    /// Returns the value in the last (most recent) checkpoint with key
+    /// lower or equal than the search key, or zero if there is none.
+    ///
+    /// This is a variant of [`Self::upper_lookup`] that is optimized to find
+    /// "recent" checkpoint (checkpoints with high keys).
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    /// * `key` - Checkpoint's key to query.
+    pub fn upper_lookup_recent(&self, key: U96) -> U160 {
+        let len = self.length();
+
+        let mut low = U256::ZERO;
+        let mut high = len;
+
+        if len > uint!(5_U256) {
+            let mid = len - len.sqrt();
+            if key < self._access(mid)._key.get() {
+                high = mid;
+            } else {
+                low = mid + uint!(1_U256);
+            }
+        }
+
+        let pos = self._upper_binary_lookup(key, low, high);
+
+        if pos == U256::ZERO {
+            U160::ZERO
+        } else {
+            self._access(pos - uint!(1_U256))._value.get()
+        }
+    }
+
+    /// Returns the value in the most recent checkpoint, or zero if
+    /// there are no checkpoints.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    pub fn latest(&self) -> U160 {
+        let pos = self.length();
+        if pos == U256::ZERO {
+            U160::ZERO
+        } else {
+            self._access(pos - uint!(1_U256))._value.get()
+        }
+    }
+
+    /// Returns whether there is a checkpoint in the structure (i.g. it
+    /// is not empty), and if so, the key and value in the most recent
+    /// checkpoint.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - read access to the checkpoint's state.
+    pub fn latest_checkpoint(&self) -> (bool, U96, U160) {
```

Comment:
Should it be `(bool, (U96, U160))`?

Reply:
Currently it is exactly like in solidity and I don't like it. I'll change it to `Option<(U96, U160)>`. That is a rust way